File: 백준/Gold/1562. 계단 수/계단 수.py
import sys


def input():
    return sys.stdin.readline().rstrip()


def solution(N: int) -> int:
    MOD = int(1e9)
    NUM_DIGIT = 10
    dp = [[[0 for i in range(1 << NUM_DIGIT)] for j in range(N + 1)] for k in range(NUM_DIGIT)]
    for digit in range(1, NUM_DIGIT):
        dp[digit][1][1 << digit] = 1

    for length in range(1, N):
        for digit in range(NUM_DIGIT):
            for mask in range(1 << NUM_DIGIT):
                val = dp[digit][length][mask]
                if 1 <= digit <= 8:
                    dp[digit + 1][length + 1][mask | 1 << digit + 1] += val
                    dp[digit - 1][length + 1][mask | 1 << digit - 1] += val
                elif digit == 0:
                    dp[1][length + 1][mask | 1 << 1] += val
                else:
                    dp[8][length + 1][mask | 1 << 8] += val
    return sum([dp[digit][N][(1 << NUM_DIGIT) - 1] for digit in range(NUM_DIGIT)]) % MOD


if __name__ == '__main__':
    print(solution(int(input())))

File: 백준/Gold/1562. 계단 수/계단 수.cc
#include <iostream>
#include <cstring>
using namespace std;
#define MOD static_cast<int>(1e9)
#define NUM_DIGIT 10
#define MAX_N 100
typedef long long ll;

ll dp[NUM_DIGIT][MAX_N + 1][1 << NUM_DIGIT];

ll getNumStair(int digit, int length, int mask) {
    if (digit == -1 || digit == 10 || length == 0 || !(mask & 1 << digit)) {
        return 0;
    }

    ll &ret = dp[digit][length][mask];
    if (ret != -1) {
        return ret;
    }
    ret = getNumStair(digit - 1, length - 1, mask)
            + getNumStair(digit - 1, length - 1, mask ^ 1 << digit)
            + getNumStair(digit + 1, length - 1, mask)
            + getNumStair(digit + 1, length - 1, mask ^ 1 << digit);
    ret %= MOD;
    return ret;
}


int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    memset(dp, -1, sizeof(dp));
    dp[0][1][1 << 0] = 0;
    for (int i = 1; i < NUM_DIGIT; i++) {
        dp[i][1][1 << i] = 1;
    }

    int N; cin >> N;
    ll ans = 0;
    for (int i = 0; i < NUM_DIGIT; i++) {
        ans += getNumStair(i, N, (1 << NUM_DIGIT) - 1);
    }
    ans %= MOD;
    cout << ans;
    return 0;
}